Working version of the multi-path-multi-projects

andreycizov opened the pull request:
There's one thing to discuss: how do we manage the destruction of the created containers?

I am leaving this up to the maintainer, but currently all containers created are assigned the same label, so that it would be easy to remove them via command line.

andreycizov committed:
Working version of the multi-path-multi-projects

hileef commented:
I like this very much, awesome work ! 😃

I have a few suggestions : 

- labelPrefix should be `fr.dldl.sphinx-server` instead of `com.dldl.sphinx-server` as `dldl.fr` is registered and owned by @quentinus95 , and `dldl.com` is not.

- I think it would be better to add logic to the containers port numbers, ( for example with something like [find-port](https://github.com/kessler/find-port) ) as right now it just starts adding up from `54334` and doesn't check if the port is already used, which could cause unfriendly errors.

Finally, I agree that we should get @Coac 's input on a container destruction/cleanup strategy.
Maybe atom provides hooks for the plugin when it closes ?

Coac reviewed (commented):
Thanks for the PR.

I agree with @hileef, the label should be `fr.dldl.sphinx-server` and using available random port would be great.

There is a problem with the current version with your changes, the container name has a trailing `=` that is not allowed by docker.
So we are getting this error 
```
(HTTP code 400) unexpected - Invalid container name (sphinx-preview-L2hvbWUvY29hYy9Qcm9qZWN0cy90ZXN0X3NwaGlueA=), only [a-zA-Z0-9][a-zA-Z0-9_.-] are allowed
```

quentinus95 commented on lib/sphinx-preview.js:
```diff
@@ -35,8 +42,44 @@ export default {
+    const subName = btoa(serverPath)
+      .replace('+', '_')
+      .replace('/', '.')
+      .replace('=', '');
```
Replace only removes the first occurence. If the base64 finishes with two equals it will fail.

quentinus95 commented on lib/sphinx-preview.js:
```diff
@@ -35,8 +42,44 @@ export default {
+  },
+
+  getContainerName(serverPath) {
+    const subName = btoa(serverPath)
```
It would be better to use a hash function here.

quentinus95 commented on lib/sphinx-preview.js:
```diff
@@ -19,6 +24,8 @@ export default {
   deactivate() {
     this.subscriptions.dispose();
+
+
```
Remove the empty lines.

andreycizov committed:
Auto port find, hashed container naming and change label namespace

andreycizov commented:
So now we use `find-free-port` to find a free port, then put it into the container label. It allows for a better interaction between separate windows of Atom: e.g. if you've got the same folder opened in two windows at the same time, they'll reuse the same docker container in order to build it.

andreycizov committed:
Fix merge conflicts with MacOS branch

andreycizov committed:
Remove debug notice

hileef commented on lib/sphinx-preview.js:
```diff
@@ -36,13 +43,54 @@ export default {
     // when docker host is macOS (darwin) Mojave
 
-    let path = `${atom.project.getPaths()[0].replace(/[/\\*]/g, '/').replace(':', '')}`;
+    let path =  `/${projectPath.replace(/[/\\*]/g, '/').replace(':', '')}`; // WINDOWS
```
```suggestion
    let path =  `${projectPath.replace(/[/\\*]/g, '/').replace(':', '')}`;
```

The leading slash for supporting windows is already added only when necessary in the lines below.
(In order to not have it when running on macOS)

hileef committed:
Update lib/sphinx-preview.js

I have no way of testing this on Windows, so hopefully someone's got access to it.

Co-Authored-By: andreycizov <acizov@gmail.com>

andreycizov committed:
Fix package-lock.json

andreycizov committed:
Merge branch 'master' of github.com:andreycizov/sphinx-preview

hileef commented:
@andreycizov , thanks for all this work 🙂 !

@Coac , @quentinus95 , do you have access to a windows machine to test this ?

@Coac , I suggest that we create a new issue and PR for the container cleanup strategy, and merge this first, if you are okay with that.

quentinus95 commented on lib/sphinx-preview.js:
```diff
@@ -36,13 +43,54 @@ export default {
+  },
+
+  getContainerName(serverPath) {
+    var hash =  new Hashes.SHA256
```
Correct me if I'm wrong, but don't we have access to NodeJS native libraries here? If so, I would prefer usage of the native library https://nodejs.org/api/crypto.html

andreycizov committed:
fix package-lock for find-free-port

andreycizov committed:
remove jshashes and use embedded crypto library
